### PR TITLE
Yaml Validator Iteration One into Main Branch

### PR DESCRIPTION
**Pull request by abaskk-msft**
## Feature
- This feature adds a Yaml Validator under the Persistence namespace. The inputs accepted are either a single yaml file or a directory of yaml files. Here is a rough UML of the structure of the project.
- Additionally users are able to see a stack trace of syntax errors. Specifically the file where an error was found and where the exact error with yaml pointer and schema pointer paths.

image

## Validation
- Tested the VSCode extension still works despite the changes in schema format
- Unit tested the Validator class

### CODE REVIEW

File: src/YAMLValidator.Tests/ValidatorTest.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Json.Schema;
+using Microsoft.PowerPlatform.PowerApps.Persistence;
```
Comment (petrochuk) [inline src/YAMLValidator.Tests/ValidatorTest.cs:5]: Please add "YamlValidatorTests" so it becomes: Microsoft.PowerPlatform.PowerApps.Persistence.YamlValidatorTests;

File: src/YAMLValidator/InputProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.CommandLine;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence;
```
Comment (petrochuk) [inline src/YAMLValidator/InputProcessor.cs:6]: Please add "YAMLValidator" so it becomes Microsoft.PowerPlatform.PowerApps.Persistence.YAMLValidator;

**Review by petrochuk — APPROVED** [review]

**Review by petrochuk — APPROVED** [review]

File: src/YAMLValidator/Constants.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Microsoft.PowerPlatform.PowerApps.Persistence.YamlValidator;
+public class Constants
+{
+    public const string FileTypeName = "file";
+    public const string FolderTypeName = "folder";
+    public const string YamlFileExtension = ".yaml";
+    public const string YmlFileExtension = ".yml";
+    public const string JsonFileExtension = ".json";
+
+    public const string verbose = "verbose";
```
Comment (tehcrashxor) [inline src/YAMLValidator/Constants.cs:13]: Style nitpick: `public const` and `public static readonly` fields should start upper case:
```suggestion
    public const string Verbose = "verbose";
```

File: docs/pa.yaml-schema.json [hunk 4]
```diff
@@ -23,7 +23,7 @@
           "type": "object",
           "properties": {
             "Control": {
-              "$ref": "control-type-schema.json#"
```
Comment (petrochuk) [inline docs/pa.yaml-schema.json:26]: This change breaks VSCode schema validation in my tests

**Review by petrochuk — CHANGES_REQUESTED** [review]

File: src/YAMLValidator/Utility.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// using YamlDotNet.Core;
```
Comment (tehcrashxor) [inline src/YAMLValidator/Utility.cs:4]: In general, we should remove commented out code
```suggestion
```

Some exceptions for work in progress, or awaiting some other change, may be fine, but should typically have an explanatory comment on why they're commented out.

File: src/YAMLValidator/YAMLValidator.csproj [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+      <NoWarn>1591, CA1822</NoWarn>
+    </PropertyGroup>
+
+    <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+      <NoWarn>1591, CA1822</NoWarn>
+    </PropertyGroup>
```
Comment (tehcrashxor) [inline src/YAMLValidator/YAMLValidator.csproj:16]: Any `<NoWarn>` in the project file should typically have a comment explaining what warnings are being suppressed.  
They're fairly simple to look up, but it's simpler for readers (either in PR or looking over the file in the future) to see it explicitly named.

**Review by petrochuk — APPROVED** [review]

File: src/YAMLValidator/InputProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        var validateCommand = new Command("validate", "Validate the input yaml file")
+        {
+            pathOption,
+            schemaOption
+        };
+
+        validateCommand.SetHandler((pathOptionVal, schemaOptionVal) =>
+        {
+            // validation has completed, we either have a file or folder
+            var pathType = File.Exists(pathOptionVal) ? Constants.FileTypeName :
```
Comment (cwduncan) [inline src/YAMLValidator/InputProcessor.cs:99]: This is a more explicit way to validate if a path leads to a folder:
`File.GetAttributes(filePath).HasFlag(FileAttributes.Directory);`

**Review by petrochuk — APPROVED** [review]

File: src/YAMLValidator/InputProcessor.cs (hunk 7 above)
Comment (abaskk-msft) [reply src/YAMLValidator/InputProcessor.cs]: added and fixed this

File: src/YAMLValidator/YAMLValidator.csproj (hunk 6 above)
Comment (abaskk-msft) [reply src/YAMLValidator/YAMLValidator.csproj]: I added an explanation as to why CA1822 was added, the other error was previously present from the Directory.Build.Props

File: src/YAMLValidator/Constants.cs (hunk 3 above)
Comment (abaskk-msft) [reply src/YAMLValidator/Constants.cs]: fixed

File: src/YAMLValidator/Utility.cs (hunk 5 above)
Comment (abaskk-msft) [reply src/YAMLValidator/Utility.cs]: just a leftover from something I tried so I ended up deleting this . I did add a comment to this based on what you suggested https://github.com/microsoft/PowerApps-Tooling/blob/afaf6d739d9152483f6040a1a9ee84d552ac0a9b/src/YamlValidator/Validator.cs#L38 makes, sense this was

**Review by tehcrashxor — APPROVED** [review]

**Pull request merged**